### PR TITLE
Quick and dirty global debug flags

### PR DESCRIPTION
**Pull request by vsoch**
This is a quick and dirty way to add support for global debug flags, currently via environment variables that then are seen by the compiler wrapper and added. It will hopefully be much more useful than #23231. We probably eventually want to make it possible to derive these envars starting from the command line, but talking with @trws it sounds like the traditional flag system / build environment variables that are set has a hook that runs after where a package can muck with the flags, meaning a user might request something and it never gets there. So we would need a different strategy.

## Questions to discuss

 - Did I add the flags in the right place (at the beginning, only because I've seen them there)?
 - How can we test this? I can test locally manually and confirm that debug symbols show up, but I'm not sure how that fits into the spack testing system

ping @trws  ! This was so much easier than the other way!

Signed-off-by: vsoch <vsoch@users.noreply.github.com>

### CODE REVIEW

**Comment by vsoch** [comment]
Ruhroh for py-setuptools!

```
  >> 1391    Bus error (core dumped)
```

**Comment by vsoch** [comment]
Another (what looks like) spurious error
```
lib/spack/spack/test/llnl/util/tty/log.py:303: AssertionError
----------------------------- Captured stderr call -----------------------------
minion: stdin.isatty(): True
minion: ready!
pid:        25498
pgid:       25498
sid:        25498
tcgetpgrp:  25498

minion pid:  25499
minion pgid: 25499
minion sid:  25498

------------------------------------------- bg
canon: on, echo: on
input: False
bg: True

!!!!!!!!!!!!!!!!!!!! Interrupted: stopping after 1 failures !!!!!!!!!!!!!!!!!!!!
```

**Comment by vsoch** [comment]
Third time is the charm! :laughing:

File: lib/spack/docs/developer_guide.rst
Comment (trws) [inline lib/spack/docs/developer_guide.rst:789]: I'm not sure what the convention is for spack (@becker33, @tgamblin?) but I think we'll want to mark this very explicitly as unstable/experimental.  We're going to want to change it, almost certainly in breaking ways, and using it without knowing what you're doing could do anything from break your build to installing a very different version of a package than what shows up in the hash.

Maybe just put "Panic" on it in big, red, friendly letters?

File: lib/spack/env/cc
Comment (trws) [inline lib/spack/env/cc]: Why g3 for preprocess mode?

File: lib/spack/env/cc
Comment (trws) [inline lib/spack/env/cc]: I like having it for documentation, but shellcheck complains about it being unused.  Maybe make it a comment?

**Review by trws — COMMENTED** [review]
This looks really good to get us started.  We might want to tweak what the selected debug options are but I think this will definitely work.  On the options I think we might just want to make it `-g` regardless of command for now, unless we want to do something more specific for when it's known to be pretending to be gcc or clang or something where we can request better data.

File: lib/spack/docs/developer_guide.rst
Comment (vsoch) [reply lib/spack/docs/developer_guide.rst:789]: One big warning section coming up!

File: lib/spack/env/cc
Comment (vsoch) [reply lib/spack/env/cc]: Changed to -g.

File: lib/spack/env/cc
Comment (vsoch) [reply lib/spack/env/cc]: Done.

**Review by trws — APPROVED** [review]
LGTM, would like at least one other set of eyes on it before merging but I'm happy.

**Review by sethrj — APPROVED** [review]
I don't work much with spack core but this looks good to me. Some thoughts:
- I don't generally like environment variables affecting builds as they can be hard to reproduce.
- Given that you can always call `strip` to remove debug symbols (like docker does) I don't think it'll be a problem

**Comment by trws** [comment]
> I don't work much with spack core but this looks good to me. Some thoughts:
> 
> * I don't generally like environment variables affecting builds as they can be hard to reproduce.
> * Given that you can always call `strip` to remove debug symbols (like docker does) I don't think it'll be a problem

FWIW, I think we all agree about environment variables, but that's the mechanism that spack uses to talk to the wrapper scripts already.  This might be used directly for now, but the intent is to have these be controlled by spack at some point in the (probably near) future.

**Comment by mwkrentel** [comment]
With this way of handling flags, does the -g flag and custom flags
feed into the package hash?  I think that's going to be important.

Another way of asking the same question is: if I run `spack find`, can
I identify which installed packages were built with -g and which were not?

So far, if SPACK_ADD_DEBUG_FLAGS is set, then all packages in the dag
will be built with -g, right?  I think that at some point you'll want
the ability to set this on a per-package basis.  This already comes
builtin for cmake.  I'm not claiming that it's easy, but I think that
eventually we will want to do that for every or most packages.

**Comment by vsoch** [comment]
@mwkrentel the quick answer is no - and this is something @trws and I discussed. We really just need a quick way to get debug flags (last week!) and it's correct that the hash won't reflect that.

For some eventual solution, I did start on https://github.com/spack/spack/pull/23231 and I met with @trws to discuss this exact PR, and we decided to start with this first. Because we need debug much sooner than that would be reasonably developed.

**Comment by sethrj** [comment]
@vsoch Since this is just "quick and dirty" are we ready to merge?

**Comment by vsoch** [comment]
I would say yes, but I don't think that's up to me :)

**Comment by sethrj** [comment]
Well, we've got approvals now, so I'm happy to hit merge unless you think this could upset anyone 😬

**Comment by vsoch** [comment]
I don't think it would upset anyone - it's a tiny change to the compile script that only adds extra debug flags given that certain envars are defined. If someone gets really angry I suppose just send them my way and I'll give them a hug and we can talk about some future implementation to do this right :)

**Pull request merged**
